yt/frontends/halo_catalog/data_structures.py

```diff
@@ -146,3 +177,363 @@ def _is_valid(self, *args, **kwargs):
+    """
+    Base class for dataset accessing particles from halo catalogs.
+    """
+
+    def __init__(self, ds, dataset_type):
+        self.real_ds = ds
+        for attr in [
+            "filename_template",
+            "file_count",
+            "particle_types_raw",
+            "particle_types",
+            "periodicity",
+        ]:
+            setattr(self, attr, getattr(self.real_ds, attr))
+
+        super(HaloDataset, self).__init__(self.real_ds.parameter_filename, dataset_type)
+
+    def print_key_parameters(self):
+        pass
+
+    def _set_derived_attrs(self):
+        pass
+
+    def _parse_parameter_file(self):
+        for attr in [
+            "cosmological_simulation",
+            "cosmology",
+            "current_redshift",
+            "current_time",
+            "dimensionality",
+            "domain_dimensions",
+            "domain_left_edge",
+            "domain_right_edge",
+            "domain_width",
+            "hubble_constant",
+            "omega_lambda",
+            "omega_matter",
+            "unique_identifier",
+        ]:
+            setattr(self, attr, getattr(self.real_ds, attr))
+
+    def set_code_units(self):
+        for unit in ["length", "time", "mass", "velocity", "magnetic", "temperature"]:
+            my_unit = "%s_unit" % unit
+            setattr(self, my_unit, getattr(self.real_ds, my_unit, None))
+        self.unit_registry = self.real_ds.unit_registry
+
+    def __repr__(self):
+        return "%s" % self.real_ds
+
+    def _setup_classes(self):
+        self.objects = []
+
+
+class YTHaloDataset(HaloDataset):
+    """
+    Dataset used for accessing member particles from YTHaloCatalogDatasets.
+    """
+
+    _index_class = YTHaloParticleIndex
+    _file_class = YTHaloCatalogFile
+    _field_info_class = YTHaloCatalogHaloFieldInfo
+
+    def __init__(self, ds, dataset_type="ythalo"):
+        super(YTHaloDataset, self).__init__(ds, dataset_type)
+
+    def _set_code_unit_attributes(self):
+        pass
+
+    @classmethod
+    def _is_valid(self, *args, **kwargs):
+        return False
+
+
+class HaloContainer(YTSelectionContainer):
+    """
+    Base class for data containers providing halo particles.
+    """
+
+    _type_name = "halo"
+    _con_args = ("ptype", "particle_identifier")
+    _skip_add = True
+    _spatial = False
+
+    def __init__(self, ptype, particle_identifier, ds=None):
+        if ptype not in ds.particle_types_raw:
+            raise RuntimeError(
+                'Possible halo types are %s, supplied "%s".'
+                % (ds.particle_types_raw, ptype)
+            )
+
+        self.ptype = ptype
+        self._current_particle_type = ptype
+        super(HaloContainer, self).__init__(ds, {})
+
+        self._set_identifiers(particle_identifier)
+
+        # Find the file that has the scalar values for this halo.
+        i_scalar = self.index._get_halo_file_indices(ptype, [self.particle_identifier])[
+            0
+        ]
```

Comment:
wow black is acting pretty weird here. Maybe the single worst example I've seen yet. Any way this line could be made a tiny bit shorter so we don't have to keep this ? (look at me complaining about the tool I advocated for, saying that it would save us future complaints about code style...)

Reply:
I don't mind leaving this as an Easter egg actually: it's pretty funny that it happens, but only because it's so rare !